## Week 11, Lab 1: SQL Practice

> Author: Matt Brems

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the famous (but fake) Northwind database. It contains various tables of transactions. The schema for the Northwind database can be found [here](http://www.zentut.com/wp-content/uploads/downloads/2013/06/Northwind-Sample-Database-Diagram.pdf).

In [3]:
import pandas as pd
from sqlalchemy import create_engine
engine = create_engine('postgres://jlpmxjkftcsepc:bb9cd75e6659d7d67ee3b86be0f905069319c75bd6823d786c808dd799ea85a2@ec2-54-227-251-33.compute-1.amazonaws.com:5432/d9teltkv30dni7')

### Prompt 1

How many rows are there in the `customers` table of the Northwind database?

In [5]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT count(*)
FROM customers;
"""

df = pd.read_sql_query(sql, engine)
df

count
0     91

**Prompt 1 Answer**:

91

### Prompt 2

If you sort the customers in alphabetical order by name, which customer is fourth?

In [99]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT contact_name
FROM customers
ORDER BY contact_name
LIMIT 1 OFFSET 3;
"""

pd.read_sql_query(sql, engine)

contact_name
0  Ana Trujillo

**Prompt 2 Answer**:

Ana Trujillo

### Prompt 3

Among those customers who live in Germany, if you sort the customers in alphabetical order by name, which customer is fourth?

In [96]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT contact_name
FROM customers
WHERE country = 'Germany'
ORDER BY contact_name
LIMIT 1 OFFSET 3;

"""

pd.read_sql_query(sql, engine)

contact_name
0  Horst Kloss

**Prompt 3 Answer**:

Horst Kloss

### Prompt 4

From the `employees` table, what are the first names of the individuals who have a birthdate in the 1960s?

In [34]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT first_name, birth_date
FROM employees
WHERE birth_date between '1960-01-01' and '1969-12-31';
"""

df = pd.read_sql_query(sql, engine)

**Prompt 4 Answer**:

In [35]:
df

first_name  birth_date
0      Janet  1963-08-30
1    Michael  1963-07-02
2     Robert  1960-05-29
3       Anne  1966-01-27

### Prompt 5

In `order_details` table of the Northwind database, what is the total number of units sold for products 2, 3, and 4? 
> You should give us three numbers - one for each `ProductID`.

In [36]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT o.product_id, sum(quantity)
FROM order_details AS o
WHERE o.product_id BETWEEN 2 and 4
GROUP By o.product_id
ORDER BY o.product_id;
"""

df = pd.read_sql_query(sql, engine)

**Prompt 5 Answer**:

In [48]:
df

product_id   sum
0           2  1057
1           3   328
2           4   453

### Prompt 6

From the `order_details` table of the Northwind database, calculate the revenue per product. In your SQL output, rename the revenue column `Revenue` and sort your results from largest revenue per product to smallest revenue per product. Which `ProductID` has the fifth-highest revenue, and what is that revenue rounded to the nearest dollar? 
> Note that revenue for a given product should be equal to the total unit price of that product times the quantity. You do not need to worry about the discount here. If you haven't multiplied columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [129]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT product_id, CAST(SUM(unit_price * quantity) AS INT) AS revenue
FROM order_details
GROUP BY product_id
ORDER BY revenue DESC
LIMIT 1 OFFSET 4;
"""
pd.read_sql_query(sql, engine)

product_id  revenue
0          62    49828

**Prompt 6 Answer**:

49828

### Prompt 7

From the `products` table of the Northwind database, show the `ProductName`, `UnitsInStock`, and `UnitsOnOrder` values for those products who have more units on order than units in stock. Among these results, which product has the most units on order?

In [60]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT product_name, units_in_stock, units_on_order
FROM products
WHERE units_in_stock < units_on_order
ORDER BY units_on_order;
;
"""

df = pd.read_sql_query(sql, engine)
df

product_name  units_in_stock  units_on_order
0         Scottish Longbreads               6              10
1               Longlife Tofu               4              20
2              Queso Cabrales              22              30
3                       Chang              17              40
4         Sir Rodney's Scones               3              40
5          Mascarpone Fabioli               9              40
6                  Gravad lax              11              50
7                    Maxilaku              10              60
8                   Chocolade              15              70
9               Aniseed Syrup              13              70
10                Rogede sild               5              70
11          Gorgonzola Telino               0              70
12  Wimmers gute Semmelknödel              22              80
13  Louisiana Hot Spiced Okra               4             100

**Prompt 7 Answer**:

In [61]:
df['product_name'][0]

'Scottish Longbreads'

### Prompt 8

From the `order_details` table, show the products that have an average discount of at least 8%. Among these results, which average discount is closest to 8%?

In [111]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT product_id, avg(discount) AS ave_disc
FROM order_details
GROUP BY product_id
HAVING avg(discount) > 0.08 
ORDER BY avg(discount);
"""

pd.read_sql_query(sql, engine)

product_id  ave_disc
0          30  0.081250
1          25  0.086111
2          58  0.100000
3           9  0.100000
4           2  0.102273
5          48  0.108333

**Prompt 8 Answer**:

In [112]:
sql = """
SELECT product_id, AVG(discount) AS ave_disc
FROM order_details
GROUP BY product_id
HAVING AVG(discount) BETWEEN 0.08 AND 0.082;
"""

pd.read_sql_query(sql, engine)

product_id  ave_disc
0          30   0.08125

### Prompt 9

Return the `orders` table, but also include the first name, last name, and country of the employee assigned to each order. What is the first and last name of the employee assigned to `OrderID` 11077? (This is the last row in the data, provided that you do not sort.)

In [113]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT o.*, e.first_name, e.last_name, e.country
FROM orders AS o
LEFT JOIN employees AS e ON o.employee_id = e.employee_id
LIMIT 10;
"""

display(pd.read_sql_query(sql, engine))

order_id customer_id  employee_id  order_date required_date shipped_date  \
0     10248       VINET            5  1996-07-04    1996-08-01   1996-07-16   
1     10249       TOMSP            6  1996-07-05    1996-08-16   1996-07-10   
2     10250       HANAR            4  1996-07-08    1996-08-05   1996-07-12   
3     10251       VICTE            3  1996-07-08    1996-08-05   1996-07-15   
4     10252       SUPRD            4  1996-07-09    1996-08-06   1996-07-11   
5     10253       HANAR            3  1996-07-10    1996-07-24   1996-07-16   
6     10254       CHOPS            5  1996-07-11    1996-08-08   1996-07-23   
7     10255       RICSU            9  1996-07-12    1996-08-09   1996-07-15   
8     10256       WELLI            3  1996-07-15    1996-08-12   1996-07-17   
9     10257       HILAA            4  1996-07-16    1996-08-13   1996-07-22   

   ship_via  freight                  ship_name  \
0         3    32.38  Vins et alcools Chevalier   
1         1    11.61         Toms Spezialitäten   
2         2    65.83              Hanari Carnes   
3         1    41.34       Victuailles en stock   
4         2    51.30           Suprêmes délices   
5         2    58.17              Hanari Carnes   
6         2    22.98          Chop-suey Chinese   
7         3   148.33         Richter Supermarkt   
8         2    13.97     Wellington Importadora   
9         3    81.91           HILARION-Abastos   

                                 ship_address       ship_city ship_region  \
0                          59 rue de l'Abbaye           Reims        None   
1                               Luisenstr. 48         Münster        None   
2                             Rua do Paço, 67  Rio de Janeiro          RJ   
3                          2, rue du Commerce            Lyon        None   
4                        Boulevard Tirou, 255       Charleroi        None   
5                             Rua do Paço, 67  Rio de Janeiro          RJ   
6                                Hauptstr. 31            Bern        None   
7                                 Starenweg 5          Genève        None   
8                          Rua do Mercado, 12         Resende          SP   
9  Carrera 22 con Ave. Carlos Soublette #8-35   San Cristóbal     Táchira   

  ship_postal_code ship_country first_name  last_name country  
0            51100       France     Steven   Buchanan      UK  
1            44087      Germany    Michael     Suyama      UK  
2        05454-876       Brazil   Margaret    Peacock     USA  
3            69004       France      Janet  Leverling     USA  
4           B-6000      Belgium   Margaret    Peacock     USA  
5        05454-876       Brazil      Janet  Leverling     USA  
6             3012  Switzerland     Steven   Buchanan      UK  
7             1204  Switzerland       Anne  Dodsworth      UK  
8        08737-363       Brazil      Janet  Leverling     USA  
9             5022    Venezuela   Margaret    Peacock     USA

**Prompt 9 Answer**:

In [110]:
sql = """
SELECT o.order_id, e.first_name, e.last_name
FROM orders AS o
LEFT JOIN employees AS e ON o.employee_id = e.employee_id
WHERE o.order_id = '11077';
"""

display(pd.read_sql_query(sql, engine))

order_id first_name last_name
0     11077      Nancy   Davolio

### Prompt 10

Find the average amount of freight across all orders, broken out by country of *employee*. (**NOT `ShipCountry`**.) How much higher is the average freight from UK employees than the average freight from US employees?

> For a bonus, round the freight to the nearest whole number and rename the column `Average Freight`.

In [131]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT AVG(o.freight) AS "average freight", e.country
FROM orders AS o
LEFT JOIN employees AS e ON e.employee_id = o.employee_id
GROUP BY e.country;
"""

pd.read_sql_query(sql, engine)

average freight country
0        78.977143      UK
1        77.973284     USA

**Prompt 10 Answer**:

In [132]:
78.977143-77.973284

1.0038589999999914

In [133]:
# Bonus
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT CAST(AVG(o.freight) AS INT) AS "average freight", e.country
FROM orders AS o
LEFT JOIN employees AS e ON e.employee_id = o.employee_id
GROUP BY e.country;
"""

pd.read_sql_query(sql, engine)

average freight country
0               79      UK
1               78     USA